plugins/unifi/unifi_clients_by_device

```diff
@@ -0,0 +1,275 @@
+my %APIconfig = (
+	"user"     => (( defined $ENV{'user'} ) ? $ENV{'user'} : "ubnt"),
+	"pass"     => (( defined $ENV{'pass'} ) ? $ENV{'pass'} : "ubnt"),
+	"host"     => (( defined $ENV{'host'} ) ? $ENV{'host'} : "https://localhost:8443"),
+	"site"     => (( defined $ENV{'site'} ) ? $ENV{'site'} : "default"),
+	"sslh"     => (( defined $ENV{'sslh'} ) ? $ENV{'sslh'} : "yes"),
+	"sslp"     => (( defined $ENV{'sslp'} ) ? $ENV{'sslp'} : "yes"),
+	"name"     => (( defined $ENV{'name'} ) ? $ENV{'name'} : "Unnamed Site"),
+);
+
+# Set up needed API endpoints and date strings
+my $APILogin  = $APIconfig{"host"} . "/api/login";
+my $APILogout = $APIconfig{"host"} . "/api/logout";
+my $APIPoint  = $APIconfig{"host"} . "/api/s/" . $APIconfig{"site"} . "/stat/device";
+my $APIData   = "{'username':'" . $APIconfig{"user"} . "', 'password':'" . $APIconfig{"pass"} . "'}";
+my $APISSLHst = ( $APIconfig{"sslh"} =~ m/no/i ) ? 0 : 2;
+my $APISSLPer = ( $APIconfig{"sslp"} =~ m/no/i ) ? 0 : 1;
+my $APIRawResp; my $APILogResp; my $ignore;
+
+# Init curl.
+my $curl = WWW::Curl::Easy->new() or die "FATAL:$me: WWW::Curl::Easy init failed!\n";
+my $retcode;
+
+# Set up curl, and login to API
+$curl->setopt($curl->CURLOPT_POST,1);
+$curl->setopt($curl->CURLOPT_COOKIEFILE,"");  # Session only cookie
+$curl->setopt($curl->CURLOPT_SSL_VERIFYPEER, $APISSLPer);
+$curl->setopt($curl->CURLOPT_SSL_VERIFYHOST, $APISSLHst);
+$curl->setopt($curl->CURL_SSLVERSION_TLSv1, 1);
+$curl->setopt($curl->CURLOPT_URL, $APILogin);
+$curl->setopt($curl->CURLOPT_POSTFIELDS, $APIData );
+$curl->setopt($curl->CURLOPT_WRITEDATA, \$APILogResp);
+$retcode = $curl->perform;
+
+if ( $retcode != 0 ) {
+	die "FATAL:$me: Unable to connect to API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+}
+
+my $jsonLogOBJ = JSON->new() or die "FATAL:$me: JSON init failed!\n";
+my $jsonLogResp = $jsonLogOBJ->allow_nonref->utf8->relaxed->decode($APILogResp);
+if ( $jsonLogResp->{"meta"}->{"rc"} ne "ok" ) {
+	die "FATAL:$me: Unable to login to API - it said: " , $jsonLogResp->{"meta"}->{"msg"} , "\n";
+}
+
+# Get some API data.
+$curl->setopt($curl->CURLOPT_HTTPGET,1);
+$curl->setopt($curl->CURLOPT_WRITEDATA, \$APIRawResp);
+$curl->setopt($curl->CURLOPT_URL, $APIPoint);
+$retcode = $curl->perform;
+
+# Decode response
+my $jsonOBJ = JSON->new() or die "FATAL:$me: JSON init failed!\n";
+my $jsonResp = $jsonOBJ->allow_nonref->utf8->relaxed->decode($APIRawResp);
+
+if ( $retcode != 0 ) {
+	die "FATAL:$me: Unable to fetch data from API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+}
+
+
+# Log back out. (Should invalidate our cookie file, just in case)
+$curl->setopt($curl->CURLOPT_HTTPGET,1);
+$curl->setopt($curl->CURLOPT_WRITEDATA, \$ignore);
+$curl->setopt($curl->CURLOPT_URL, $APILogout);
+$retcode = $curl->perform;
+
+if ( $retcode != 0 ) {
+	die "FATAL:$me: Unable to logout of API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+}
+
```

Comment:
All of your plugins share the above ~200 lines of code and comments (with slight differences).
It would be an amazing reduction of the maintenance burden for these upcoming plugins, if you would merge them into one single plugin providing many multigraph.

Reply:
> [..] but munin-graph, assuming it's not running via CGI, would be painful.

Starting with munin 3.0 only CGI mode (as it is called for 2.0) is available. Thus if your concern only relates to CGI/cron mode, then you may feel free to assume CGI mode to be the default.

Regarding performance / resource usage: turning the one-for-all plugin into a wildcard plugin could be the solution for selecting a specific subset of the plugin's features. But I could imagine, that most users will want to use *all* features of the plugin, or? In this case I would suggest to use a single plugin without the wildcard feature.

> [..] what, if any work has been done on explaining best case scenarios where "suggest" can't work before some basic configuration is done? Is this even a possible workflow?

I am not sure, that I understand your question fully. But I will try to answer it :)
There are two types of wildcard plugins:
* A) a plugin that allows the selection of different topics/aspects (e.g. your plugins)
* B) a plugin that can be applied to different targets (e.g. `if_eth0`)

A) would simply return a static list of keywords during `suggest`.
B) would try to discover the list of acceptable targets dynamically.

Does this answer your question?

Reply:
Cool - that makes sense.  I think what I'll do is go monolithic - in the same manner that I did with unifi_device_stats - everything on be default, but allow individual aspects to be shut off.  It's more config, but as you said, I would imagine the vast majority of users would want all of the bits turned on.

And thats good to hear on the CGI front, that does remove my concern about performance totally - actually, the single plugin approach will actually allow me to reduce the number of API calls, which is also not a bad thing at all - right now, it hits the API 3 times per plugin, and the unifi API is not the fastest thing to query over the network - it's not bad, but it definitely had an impact on the runtime of munin-update.  Thanks for the help, expect to see some of these updates added to this pull request sometime soon - it's kinda been nice, as I hadn't touched perl in *years*, working the brain a bit hard.